cmd/proxy/actions/app_proxy.go

```diff
@@ -46,20 +46,28 @@ func addProxyRoutes(
 	// 3. The stashpool manages limiting concurrent requests and passes them to stash.
 	// 4. The plain stash.New just takes a request from upstream and saves it into storage.
 	fs := afero.NewOsFs()
-	mf, err := module.NewGoGetFetcher(goBin, fs)
+	mf, err := module.NewGoGetFetcher(c.GoBinary, fs)
 	if err != nil {
 		return err
 	}
 
-	lister := download.NewVCSLister(goBin, fs)
-	st := stash.New(mf, s, stash.WithPool(goGetWorkers), stash.WithSingleflight)
+	lister := download.NewVCSLister(c.GoBinary, fs)
+	sf := stash.WithSingleflight
+	if c.EtcdEndpoints != "" {
```

Comment:
Probably we  should think of some configuration variable to select `stasher` (probably more of them to come). Something like we do with `storages` . It probably could be done in follow up PR.

Reply:
@t-tomalak we should think more about this. The less the user has to configure the better.

Reply:
i think @t-tomalak had multiple providers in mind. for now we have etcd but we cant force everybody to use it, similar structure as with storages would be much more valuable

Reply:
can you also encapsulate into factory func maybe? as singleflight is in memory sync and etcd is distributed sync we can just call a func which returns `sf,err`